Don't hold ThreadPool lock when destructing task

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57030 Assert that GIL is not held in blocking destructors
* #57029 Release GIL before destructing RPCAgent subclasses.
* **#56817 Don't hold ThreadPool lock when destructing task**

Fix https://github.com/pytorch/pytorch/issues/56701 and https://github.com/pytorch/pytorch/issues/56786

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27975642](https://our.internmc.facebook.com/intern/diff/D27975642)

ezyang committed:
Don't hold ThreadPool lock when destructing task

Fix https://github.com/pytorch/pytorch/issues/56701 and https://github.com/pytorch/pytorch/issues/56786

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

pritamdamania87 commented on c10/core/thread_pool.cpp:
```diff
@@ -114,21 +114,22 @@ void ThreadPool::main_loop(std::size_t index) {
-        complete_ = true;
-        completed_.notify_one();
-      }
+    } // Destruct tasks before taking the lock
```
nit: would be useful to add more details here about the potential deadlock if this is not done.

ezyang committed:
Update on "Don't hold ThreadPool lock when destructing task"


Fix https://github.com/pytorch/pytorch/issues/56701 and https://github.com/pytorch/pytorch/issues/56786

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27975642](https://our.internmc.facebook.com/intern/diff/D27975642)

[ghstack-poisoned]

ezyang committed:
Update on "Don't hold ThreadPool lock when destructing task"


Fix https://github.com/pytorch/pytorch/issues/56701 and https://github.com/pytorch/pytorch/issues/56786

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27975642](https://our.internmc.facebook.com/intern/diff/D27975642)

[ghstack-poisoned]
